tests/test_metadata.py

```diff
@@ -0,0 +1,413 @@
+# This file is dual licensed under the terms of the Apache License, Version
+# 2.0, and the BSD License. See the LICENSE file in the root of this repository
+# for complete details.
+
+import dataclasses
+import json
+import tarfile
+from email.policy import compat32
+from functools import partial
+from hashlib import md5
+from itertools import chain
+from pathlib import Path
+from textwrap import dedent
+from typing import Iterator, List
+from urllib.request import urlopen
+from zipfile import ZipFile
+
+import pytest
+
+from packaging.metadata import (
+    CoreMetadata,
+    InvalidCoreMetadataField,
+    InvalidDynamicField,
+    MissingRequiredFields,
+    UnfilledDynamicFields,
+)
+from packaging.requirements import Requirement
+from packaging.utils import canonicalize_name
+from packaging.version import Version
+
+HERE = Path(__file__).parent
+EXAMPLES = HERE / "metadata_examples.csv"
+DOWNLOADS = HERE / "downloads"
+
+
+class TestCoreMetadata:
+    def test_simple(self):
+        example = {
+            "name": "simple",
+            "version": Version("0.1"),
+            "requires_dist": [Requirement("appdirs>1.2")],
+        }
+        metadata = CoreMetadata(**example)
+        assert isinstance(metadata.to_pkg_info(), bytes)
+
+    def test_invalid(self):
+        example = {
+            "name": "simple",
+            "author_email": [(None, "me@example.com")],
+            "requires_dist": [Requirement("appdirs>1.2")],
+        }
+        metadata = CoreMetadata(**example)
+        with pytest.raises(MissingRequiredFields):  # version is missing
+            metadata.to_pkg_info()
+
+        metadata = dataclasses.replace(metadata, version=Version("0.42"))
+        with pytest.raises(InvalidCoreMetadataField):
+            dataclasses.replace(metadata, dynamic=["myfield"]).to_pkg_info()
+        with pytest.raises(InvalidDynamicField):
+            dataclasses.replace(metadata, dynamic=["name"]).to_pkg_info()
+
+    PER_VERSION_EXAMPLES = {
+        "1.1": {
+            "has_dynamic_fields": False,
+            "is_final_metadata": True,
+            "file_contents": """\
```

Comment:
Do we want to embed the file contents, or should we have them as files in the test suite that we load as appropriate?